### PR TITLE
chore: cleanup nix!

### PR DESCRIPTION
**Pull request by ajbt200128**
Added some comments explaining more things, cleaned up the flake a bit, fixed flags.sh to actually fix the nix flake, then setup caching!

## Test plan
```bash
# with nix installed
make nix-check
make shell
make core
```
CI step "build-test-nix" passes

### CODE REVIEW

File: src/main/flags.sh [hunk 1]
```diff
@@ -49,7 +49,16 @@ OS="$1"
 # - we're on alpine, in which case the opam switch doesn't have a special
 #   name. It is assumed that the reason we're on alpine is to get
 #   statically-linked executables.
-if [[ "$(opam switch show)" == *+static* ]]; then
+
+
+# Check if NIX_ENVIRONMENT is not set
+# If it is set, we are in a nix-shell and we should not statically link
+# This is first because opam won't exist in a nix build environment
+if [[ -n "${SEMGREP_NIX_BUILD-}" ]]; then
```
Comment (kopecs) [inline src/main/flags.sh:57]:
```suggestion
if [[ ! -v SEMGREP_NIX_BUILD ]]; then
```
can we just use `-v` here for consistency given we require bash 5+ anyway?

Comment (kopecs) [reply src/main/flags.sh:57]: (untested)

**Review by kopecs — APPROVED** [review]

File: src/main/flags.sh (hunk 1 above)
Comment (ajbt200128) [reply src/main/flags.sh:57]: trying

Comment (ajbt200128) [reply src/main/flags.sh:57]: yea GHA uses bash 4.2 or smthn, workflows failed, force pushing back down to prev commit

File: .github/workflows/build-test-nix.jsonnet [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        name: "Cache Shells and Binaries",
+        run: "make nix-cache"
+    }
+
+    */
+]);
+{
+  name: 'build-test-nix',
+  // This is called from tests.jsonnet and release.jsonnet
```
Comment (aryx) [inline .github/workflows/build-test-nix.jsonnet:51]: remove this comment. We are not called from tests.jsonnet actually here.

File: .github/workflows/build-test-nix.jsonnet [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+local job = gha.os_matrix(
+  oss=['ubuntu-latest', 'macos-latest'],
+  steps=[
+    gha.speedy_checkout_step,
+    actions.checkout_with_submodules(),
+    {
+        name: "Set up Nix",
+        uses: "DeterminateSystems/nix-installer-action@main",
+        with: {
+            "extra-conf": "sandbox = false"
```
Comment (aryx) [inline .github/workflows/build-test-nix.jsonnet:17]: would be good to comment why we need this sandbox = false. What happens if we put true? What is failing?

File: Makefile [hunk 4]
```diff
@@ -462,7 +460,51 @@ homebrew-setup:
 # -------------------------------------------------
 # Nix
 # -------------------------------------------------
-#TODO: see flake.nix
+# See flake.nix top level comments for more information
+
+# Enter development environment with all dependencies installed
+shell:
+	$(eval USER_SHELL := $(shell finger ${USER} | grep 'Shell:*' | cut -f3 -d ":"))
```
Comment (aryx) [inline Makefile:467]: what is this thing?

File: Makefile [hunk 5]
```diff
@@ -462,7 +460,51 @@ homebrew-setup:
...
+	nix develop -c $(USER_SHELL)
+
+# Build targets
+nix-osemgrep:
+	nix build ".?submodules=1#osemgrep"
```
Comment (aryx) [inline Makefile:472]: what is this magic .?submodules thing?

**Review by aryx — COMMENTED** [review]
ok

File: .github/workflows/build-test-nix.jsonnet (hunk 3 above)
Comment (ajbt200128) [reply .github/workflows/build-test-nix.jsonnet:17]: will add a comment

File: .github/workflows/build-test-nix.jsonnet (hunk 2 above)
Comment (ajbt200128) [reply .github/workflows/build-test-nix.jsonnet:51]: will add a comment

File: Makefile (hunk 4 above)
Comment (ajbt200128) [reply Makefile:467]: will add a comment, but it gets the user shell

**Pull request merged**